### PR TITLE
Extend type_string with wait_still_screen parameter

### PR DESCRIPTION
**Pull request by yxususe**
A solution for: https://progress.opensuse.org/issues/19872
Testrun: http://f146.suse.de/tests/1045

### CODE REVIEW

**Comment by okurz** [comment]
makes sense to me but unfortunately does not work. See the output in autoinst-log.txt:

```
15:03:50.8242 12171 >>> testapi::_handle_found_needle: found root-console-20170324, similarity 1.00 @ 0/1
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
15:03:50.8243 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:7 called testapi::type_string
15:03:50.8244 12171 <<< testapi::type_string(string='Do not wait_still_screen', max_interval=250, wait_screen_changes=0)
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
15:03:51.4152 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:8 called testapi::type_string
15:03:51.4154 12171 <<< testapi::type_string(string='wait_still_screen for 5 seconds', max_interval=250, wait_screen_changes=0)
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
15:03:52.1797 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:9 called testapi::type_string
15:03:52.1798 12171 <<< testapi::type_string(string='wait_still_screen for 10 seconds', max_interval=250, wait_screen_changes=0)
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
15:03:53.0031 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:10 called testapi::type_string
15:03:53.0032 12171 <<< testapi::type_string(string='wait_still_screen for 20 seconds', max_interval=250, wait_screen_changes=0)
```

So there are error messages, the log output does not mention the arguments of wait_still_screen and also in the time stamp one can see that there is no difference.
Are you sure you tested the version of os-autoinst that you changed? You would need to build os-autoinst locally (`./autogen.sh && ./configure && make`) and then call an openQA worker with the argument for a custom location of os-autoinst, e.g. `/usr/share/openqa/script/worker --isotovideo /home/yxu/os-autoinst/isotovideo --instance 1`. What might be easier and appreciated is extending the unit tests of os-autoinst instead. Take a look in t/03-testapi.t for that and call `make check` to execute.

Additionally, please tidy (should fix the travis check) and delete the dot at end of the git commit subject line.

**Comment by yxususe** [comment]
Hi Oliver, can you explain a bit more about the following:
What might be easier and appreciated is extending the unit tests of os-autoinst instead. Take a look in t/03-testapi.t

**Comment by okurz** [comment]
discussed in jangouts call, better to adress this some days later in person

**Comment by yxususe** [comment]
New testrun didn't work yet. 
http://f146.suse.de//tests/1058

**Comment by okurz** [comment]
are you sure you called the custom worker? See https://github.com/os-autoinst/os-autoinst/pull/842#issuecomment-324060957

**Comment by yxususe** [comment]
Yes, it called. 
> 
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
09:34:06.4019 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:7 called testapi::type_string
09:34:06.4021 20020 <<< testapi::type_string(string='Do not wait_still_screen', max_interval=250, wait_screen_changes=0)
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
09:34:06.9978 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:8 called testapi::type_string
09:34:06.9979 20020 <<< testapi::type_string(string='wait_still_screen for 5 seconds', max_interval=250, wait_screen_changes=0)
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
09:34:07.7949 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:9 called testapi::type_string
09:34:07.7951 20020 <<< testapi::type_string(string='wait_still_screen for 10 seconds', max_interval=250, wait_screen_changes=0)
Odd number of elements in hash assignment at /usr/lib/os-autoinst/testapi.pm line 1081.
09:34:08.5970 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:10 called testapi::type_string
09:34:08.5972 20020 <<< testapi::type_string(string='wait_still_screen for 20 seconds', max_interval=250, wait_screen_changes=0)

**Comment by yxususe** [comment]
A new testrun: http://f146.suse.de/tests/1067
Log shows the parameter for wait_still_screen didn't work:
15:16:33.0631 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:22 called testapi::type_string
15:16:33.0632 26932 <<< testapi::type_string(string='wait_still_screen for 20 seconds
', max_interval=125, wait_screen_changes=0)
15:16:36.8601 Debug: /var/lib/openqa/share/tests/sle/tests/console/type_string_test.pm:23 called testapi::type_string
15:16:36.8606 26932 <<< testapi::type_string(string='date
', max_interval=250, wait_screen_changes=0)

**Comment by yxususe** [comment]
Newest test result can be seen here:
https://w3.suse.de/~yxu/autoinst-log.txt
The logic part between wait_still_screen and wait_screen_change in testapi.pm still needs to be improved. Besides that, parameter wait_still_screen already works, and max_interval also works well.

**Comment by okurz** [comment]
LGTM except for conflicts in t/99-full-stack.t which need to be resolved by rebasing

**Comment by yxususe** [comment]
WIP updated. New logfile available here: https://w3.suse.de/~yxu/autoinst-log.txt

**Comment by yxususe** [comment]
Most recent logfile available at: https://w3.suse.de/~yxu/autoinst-log.txt

**Comment by yxususe** [comment]
Hi Santi, about the test of hello, I have already removed them, now the tests are the following 5 lines:
```
    type_string ("echo hello\n", max_interval => 1, wait_still_screen => 2);
    type_string ("echo do not wait_still_screen\n", max_interval => 1, wait_still_screen => 0);
    type_string ("echo test\necho wait\necho 10se\n", max_interval => 100, wait_screen_changes => 11, wait_still_screen => 10);
    type_string ("echo test if wait_screen_change functions as expected\n", max_interval => 100, wait_screen_changes => 11, wait_still_screen => 10);
    type_string ("echo wait_still_screen for 20 seconds\n", max_interval => 200, wait_still_screen => 20);
```

**Comment by yxususe** [comment]
PR updated.

File: t/03-testapi.t [hunk 1]
```diff
@@ -109,6 +109,17 @@ type_string 'hallo', wait_screen_change => 3, max_interval => 10;
 is_deeply($cmds, [{cmd => 'backend_type_string', max_interval => 10, text => 'hal'}, {cmd => 'backend_type_string', max_interval => 10, text => 'lo'},]);
 $cmds = [];
 
+subtest 'type_string with wait_still_screen' => sub {
+    my $wait_still_screen_called = 0;
+    my $module = new Test::MockModule('testapi');
```
Comment (foursixnine) [inline t/03-testapi.t]: Use tools/tidy, looks like your code is failing https://travis-ci.org/os-autoinst/os-autoinst/builds/277241064#L1728

**Comment by yxususe** [comment]
PR updated.

File: testapi.pm [hunk 2]
```diff
@@ -1123,17 +1123,24 @@ sub type_string {
 
     my $max_interval = $args{max_interval}       // 250;
     my $wait         = $args{wait_screen_change} // 0;
-    bmwqemu::log_call(string => $log, max_interval => $max_interval, wait_screen_changes => $wait);
+    my $wait_still   = $args{wait_still_screen}  // 0;
```
Comment (okurz) [inline testapi.pm]: Please update the API description in the documentation above

Comment (foursixnine) [reply testapi.pm]: @okurz Good catch :)... /cc @yxususe

**Pull request merged**